File: problems/LeetCode/TopInterview150/canConstruct.ts
function canConstruct(ransomNote: string, magazine: string): boolean {
  let noteObj = {} as Record<string, number>;
  let magObj = {} as Record<string, number>;

  for (let char of ransomNote) {
    if (!noteObj.hasOwnProperty(char)) {
      noteObj[char] = 1;
    } else {
      noteObj[char]++;
    }
  }

  for (let char of magazine) {
    if (!magObj.hasOwnProperty(char)) {
      magObj[char] = 1;
    } else {
      magObj[char]++;
    }
  }

  for (let char of ransomNote) {
    if (!magObj.hasOwnProperty(char) || magObj[char] <= 0) {
      return false;
    }

    magObj[char]--;
  }

  return true;
}

// Time complexity: O(n)
// Space completiry: O(n)

function canConstruct2(ransomNote: string, magazine: string): boolean {
  if (ransomNote.length > magazine.length) {
    return false;
  }

  for (let char of new Set(ransomNote)) {
    if (magazine.split(char).length - 1 < ransomNote.split(char).length - 1) {
      return false;
    }
  }

  return true;
}

canConstruct('a', 'b');

File: problems/LeetCode/TopInterview150/canConstruct.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:

        count = {}

        for char in magazine:
            count[char] = count.get(char, 0) + 1

        for char in ransomNote:
            if count.get(char, 0) == 0:
                return False

            count[char] -= 1

        return True

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:

        # keeps only characters that are still missing. For example:
        # ransomNote = "aa"
        # magazine = "ab"
        # Counter(ransomNote) - Counter(magazine)
        # Counter({'a': 1})
        # Because the result is not empty, not converts it to False.

        return not Counter(ransomNote) - Counter(magazine)


# Time: O(n)
# Space: O(n)